isearch拡張機能のadviceによる再実装

tooro88 opened the pull request:
「emacs 本体の関数の上書きを無くしたい #29」に書いた、advice による isearch 拡張と wrapped-search の実装です。テストも付けました。input method 系の実装は、この commit に追加する形で後日、別 PR にします。

以下、commit log に書ききれなかったコメントです。

#### Reimplement isearch extensions using advice (commit e8391d7)

 - 実装本体です。「isearch-lax-whitespaceが機能しない #25」も自動で直っています。
 - tc-isbushumaze.el の関数は tc-is22.el の部首/交ぜ書き変換関連コードのコピーで、1文字も変えていません。
    + tc-is22.el がこのまま存続し続けるなら、ts-is22.el からもこのファイルをロードする形にするのがよいと思います。
 - tc-ishelper.el の copyright 部分は適当に書いたので確認お願いします。
 - wrapped search の実装には、isearch.el 内の `isearch-define-mode-toggle` というマクロを使いました。これには、「キーバインドを作る」という望んでない機能もついていて、オフにできないのですが、まあいっかと思ってそのままにしてあります。iseach 中に `M-s @` で wrapped search がオンオフできます。
 - `tcode-isearch-start-state` については、#28 の問題を修正した版を入れました。この機能の発案者がどういう意図だったのか(「バッファと独立」とは?)、元の実装がその意図に沿ったものだったのか、いまいちよくわからないのですが、元の実装の挙動を仕様と考えて、DOCSTRING を明確化しました。
    + 値 `t` は無くしてあります。

#### Add a section about isearch extensions to README.md (commit b9a73bc)
 - isearch 実装が選べることを README.md に書きました。

#### Add ert tests for isearch extensions (commit f2c9ab5)
 - テストです。使い方は tctest.el をご覧ください。
 - 将来の input method 方式のテストも含めてあります。そこで改善される問題(多段前置部首変換が可能になるなど)のテストは失敗しますが、既知の問題扱い(`:expected-result` が `:fail`) としてあるので、テスト全体では成功扱いになるはずです。(赤いFマークが出ない)。
 - 以下の設定で、Windows 版 emacs の24〜30の各メジャーバージョンでテストが通る(unexpected な結果が出ない)ことを確認しました。
    + `(setq tcode-use-isearch nil)`
    + `(setq tcode-use-isearch t)`
    + `(setq tcode-use-isearch :advice)`
 - tctest-env.el、run.bash は、自動実行を楽にするためのスクリプトです。各自でやればよい話とも言えますが、せっかく作ったので入れておきました。

以上です。修正点、疑問点等あればお知らせください。些細な問題でも教えていただけるとありがたいです。

naota commented:
> ### commit 分割
> 目的ごとに分割されている方が読みやすいとのことであれば、advice 化、部首/交ぜ書きサポート、wrapped search はそれぞれ独立した修正なので、わける方が各 commit がすっきりすると思います。次のようになるでしょうか。
> 
> * `tcode-isearch-start-state` の tc-is22.el 上での修正。
> * tc-is22.el の分割。tc-iscommon.el を作る。
> * 機能選択と空の ishelper.el のロード部分 (こちらの commit が先にあった方が続く commit のテストがしやすいので)
> * advice による isearch 実装
> * tc-iscommon.el 読み込みによる、部首/交ぜ書き変換、`tc-isearch-start-state` のサポート
> * wrapped search 実装

よいと思います. 細かい方が目的がはっきりしてレビューの方向が定まりますし, ここまでOKという区切りもつけやすくて助かります.

tooro88 commented:
commit を作り直しました。よく考えたら、「advice による isearch 実装」と「部首/交ぜ書き変換」は分けられないのでした。その2つはまとめてあります。次のようになりました。
 - (commit f4237f3) `tcode-isearch-start-state` の tc-is22.el 上での修正
 - (commit bd85e06) tc-is22.el の分割
 - (commit 68fbc1f) 機能選択と空の ishelper.el のロード部分
 - (commit cf898eb) adviceによるisearch 実装、部首/交ぜ書き変換と `tcode-isearch-start-state` 修正の有効化
 - (commit 9622ce3) wrapped search 実装
 - (commit 7a38637) README.md への記述追加
 - (commit 31fafbb) テスト作成

naota commented on tc-pre.el:
```diff
@@ -46,7 +46,7 @@
 
 (defconst tcode-isearch-type 'tc-is22
-  "isearchで用いるTコード用モジュールのタイプ。")
+  "tcode-use-isearch が t のとき、isearch で用いるモジュール名。")
```
他でも書きましたが, ここをdefconstではなく, defcustomにして, tc-is22やtc-ishelperの値をとればいいかなと思いました.

tooro88 replied:
了解です。確かに `t` より `overwrite` の方がメッセージ性があってわかりやすいと思います。`tcode-use-isearch` の値を `overwrite`、`advice`、`nil`、(後に `im` も) としつつ、従来の値 `t` も許容する(`overwrite` とみなす)ように、commit を修正します。

`tcode-isearch-type` を、`tcode-overwrite-module` に変更する commit も別途追加します。

ファイルは分割するほど巨大というわけでもないので、ファイル構成は一旦このままで行かせてください。あとで「やっぱり変だ」となったらお知らせください。

naota commented on tc-setup.el:
```diff
@@ -24,15 +24,25 @@
+	 ;; .tc を用意しているユーザーへの default 値。
+	 ;; nil に設定したい場合は、.tc でセットする必要がある。
+	 (setq tcode-use-isearch (or tcode-use-isearch t))
+	 (setq tcode-use-as-default-input-method t)
```
このsetqもよくわからないです. tc-ishelperの動作に必要ということでしょうか?

tooro88 commented on tc-is22.el:
```diff
@@ -31,15 +31,12 @@
 ;;;  User Variables
 ;;;
-(defvar tcode-isearch-start-state nil
+(defvar-local tcode-isearch-start-state nil
```
はい、削除 commit を追加します。

tooro88 commented on tc-pre.el:
```diff
@@ -72,7 +72,14 @@ mule-1, mule-2, mule-3, mule-4, xemacsのいずれか。")
+ t       : 従来の tc-is22.el による実装を用いる。
+ \\='advice : advice を使うことにより Emacs 内部関数の書き換えを無くした
+           実装を用いる。
+デフォルト値は t。nil にするには、設定ファイル ~/.tc 内でセットする。")
```
これ、現状の仕様(挙動?)は、設定ファイル ~/.tc を持っていない人のデフォルト値が `nil`、持っている人のデフォルト値が `t` という風になっています。今回、`advice` の説明を追加したついでにデフォルト値に関する説明を足しましたが、.tc を持っていない人が tcode 変数の Help を読むとも思えないので、説明を簡略化しています。必要なら詳しく書き直します(またはデフォルト値への言及をやめてもよいと思います)。

この仕様自体、意味のある仕様とも思えないので、やめてしまう(全員 `overwrite` をデフォルトにする)という手もありますが、今このタイミングで手を付けたいとは思いません。実装についてはこのままとさせてください。(`t` を `overwrite` に変える変更だけします。)

naota commented on tc-ishelper.el:
```diff
@@ -80,6 +80,58 @@
 (defun tcode-isearch-make-string-for-wrapping (s) s)
 
+;;;
+;;; wrapped-search
```
そうか…もともと tcode-isearch-enable-wrapped-search の名前があるんですね。むずかしいなあ…

naota replied:
newline foldedか、newline incensitiveとかですかねえ…。ちなみにmigemoの方を見てみましたが、特に名前はつかず、defaultでただそのようになる感じでした。

naota replied:
あ、line foldでいきましょう

tooro88 replied:
はい、では、あちこちの wrapped search と書かれている部分を、line-fold search に直します。
昔からある `tcode-isearch-enable-wrapped-search` という変数については、`define-obsolete-variable-alias` というちょうどよい関数があるようなので、それで alias に設定しておきます。

tooro88 commented on tc-ishelper.el:
```diff
@@ -80,6 +80,58 @@
+;; isearch 内のコマンド isearch-toggle-wrap を作り、"M-s @" にバインドする。
+;;  - キーバインドは特に必要ないが、このマクロの仕様上スキップできない。
+;;    一時的にバインドしてすぐに元に戻す。
+;;  - ???: このマクロはトップレベルに置かないとエラー。emacs -qでこの
```
私も再現方法を忘れていたのですが、何とか再現し直すことができました。emacs-25 か 26 のみで起きる問題でした。次の内容のファイルを `emacs -Q err.el` として読み込んで、`M-x eval-buffer` とします。(`if` の条件部分は何でもよいです。)

~~~elisp
(defun const-a-regexp (&rest args) "a")
(if (fboundp 'isearch-define-mode-toggle)
    (progn
      (require 'cl-macs)
      (isearch-define-mode-toggle const-a "@" const-a-regexp)))
~~~

emacs-27 以降では、このあと、`C-s M-s @ foo` とすると、`foo` の部分を何と入力しても、`a` をサーチします(意図通りの動作)。emacs-25, 26 では、`eval-buffer` の時点で次のエラーになります。

~~~elisp
Invalid function: (function (lambda (types) (cons (quote choice) (cons (quote (const :tag "Const-A Search" const-a-regexp)) (cdr types)))))
~~~

今試したところ、`(require 'cl-macs)` の方だけトップレベルに置けば正しく動作するようです。これも原因はよくわかりません。

tooro88 replied:
トップレベルに置くなら、`eval-when-compile` なしで `(require 'cl-macs)` だけでも正しく動くようです。if文の中に `(require 'cl-macs)` を置いた場合は、`(eval-when-compile ...` を付けると、正しく動くようになりました。

どうもロード時のマクロ展開と実行時のマクロ展開で、内部で使っている `cl-callf` の展開のされ方が違うような感じなのですが、詳しくはよくわかりません。古いバージョンでだけ起こるバグっぽい挙動なので、あまり調べても仕方ないかなと思ってます。

ところでこの問題とは別の話として、elisp の info によると、マクロ定義が目的で何かを `require` する場合は、`eval-when-compile` で囲むのが正しいようですね(`require` している側をコンパイルした .elc のロード時に、不要な `require` が発生しなくなるので)。そのように直しておきます。

naota commented on tc-ishelper.el:
```diff
@@ -80,6 +80,58 @@
+  (when (and isearch-mode
+	     tcode-isearch-enable-wrapped-search
+	     (boundp 'isearch-regexp-function) ; この2行は emacs-24 での
+	     (fboundp 'isearch-toggle-wrap)    ; compiler warning 対策。
```
ああ、そうですね。これもいれましょう。このREADMEのpatchはこのPRとは独立してささっと入れてもいいですね。

tooro88 replied:
はい、PR #35 を作成しました。

tooro88 commented:
レビュー結果を反映しました。テストのコードも少し整理と機能追加しています。

変更した commit は以下の通りです。
 - (commit a07d392) 変更なし
 - (commit 21269c7) 変更なし
 - (commit 978963d) 変更なし
 - (commit f005467) 変更なし
 - (commit b49f104) 変更なし
 - (commit 872ddd4) Add 'advice option to tcode-use-isearch
    + `tcode-use-isearch` で存在しない値が使われた場合に警告を出すようにした。
 - (commit ef228c5) Reimplement Isearch extensions using advice
    + `tcode-isearch-make-string-for-wrapping` の定義を `defalias` に。
 - (commit 4f20c1a) Reimplement wrapped search using isearch-regexp-function
    + emacs-24 用の場合分けを無くした。
    + `require` を `eval-when-compile` で囲んだ。
    + `isearch-define-mode-toggle` 前のキー定義の待避を `let` で行なうようにした
 - (commit 57e1816) Add a section about Isearch extensions to README.md
    + emacs-24 への言及を削除。
 - (commit 44eb884) Add ERT tests for Isearch extensions
    + emacs-24、25 用の場合分けを無くした。
    + `tcode-use-as-default-input-method` を `nil` に設定して起動できるようにした。
    + テスト用のキー定義コードを整理した。

naota commented on tc-pre.el:
```diff
@@ -72,7 +73,15 @@ mule-1, mule-2, mule-3, mule-4, xemacsのいずれか。")
+ nil         : isearch 中にTコードを使用しない。
+ `overwrite' : Emacs 内部関数の書き換えによる実装(従来実装)を用いる。
+ `advice'    : advice を使うことにより Emacs 内部関数の書き換えを
+               無くした実装を用いる。
```
ちょっと細かくなりますが、「書き換えをなくした」というのは(ユーザにはあまり関係のない)経緯的でドキュメントとしてはそぐわないかも?と思いました。
「advice を使う実装」や「advice を使い、 Emacs 内部関数を書きかえない実装」(経緯的意図を省く)というのはいかがでしょう。
まあ、ここはだいぶ好みなので、どれでも大丈夫だとは思います。

tooro88 replied:
o なるほど、確かに。そもそもこの文はくどくて、advice を知らない人がいるかもと考えてこうしてしまったのですが、上の行と対比して書かれているのだから、内部関数の書き換えをしないことは十分伝わりますよね。`im` 版も含めて、こんな感じにしようと思います。

```elisp
(defvar tcode-use-isearch nil
  "nil でないとき、Tコードを使用できるように isearch を拡張する。

次のシンボルのいずれかを指定する。
 nil         : isearch 中にTコードを使用しない。
 `overwrite' : Emacs 内部関数の書き換えによる実装(従来実装)を用いる。
 `advice'    : advice による実装を用いる。
 `im'        : input method による実装を用いる。
               `tcode-use-input-method' が自動で t となる。
デフォルト値は `overwrite'。nil にするには、設定ファイル ~/.tc 内で
セットする。")
```

naota commented on tc-setup.el:
```diff
@@ -24,15 +24,25 @@
+  (cond ((memq tcode-use-isearch '(t overwrite))
+         (require 'tc-sysdep)
+         (require tcode-isearch-overwrite-module))
+        (tcode-use-isearch
```
あれ、ここって更新されてますか?

naota replied:
あ、tcode--ensure-valid-value()の部分でsetしてるのですね…なるほど

naota commented on tc-setup.el:
```diff
@@ -24,15 +24,36 @@
+        (tcode-use-isearch
+         (require 'tc-ishelper))))
+
+(defun tcode--ensure-valid-value (var vals fallback)
```
この関数は後で使われたりしますか? 関数内で変数がsetされる副作用があって、効果が関数型的でない点、condでの場合分けでも十分に書ける点を考えると、他での用途がなければ、condでそのまま書きたいかと思います。

tooro88 replied:
`tcode--ensure-valid-value` は、現状予定はないものの、将来別の変数で使うことも想定して書いたものです。`tcode-use-isearch` 一つでの使用に限ったとしても、論理的には複数個所から使うかも知れない処理(`tcode-use-isearch` の最初の使用の前に置きたい。たまたま最初の使用場所がこの一箇所に限られるだけ)なので、関数になっている方が自然だと感じます。

(仮に論理的にも1ヶ所でしか使われない関数だとしても、独立した意味のあるまとまりとして名前の付けられる操作は、どんどん関数にすべき、それが良いプログラムであると私は考えます。)

副作用が見えにくいことを問題視しているのであれば、こう考えてもらうことはできないでしょうか? もともと `tcode-use-isearch` は `nil` `t` `overwrite` `advice` の4値しか取らない変数で、そうでない値が来たときはバグです。そのようなバグが無いことを保証するのが `tcode--ensure-valid-value` です。`assert` 文のようなものです。もし `assert` 文のように、`tcode--ensure-valid-value` の中で、警告の代わりにエラーにしていたら、問題視しないのではないでしょうか。だとしたら、現行版はエラー版でやりたいことと同等で、より親切な挙動になっていると考えられないでしょうか。

naota replied:
ここはかなり悩ましいのですが、 `assert` が成り立つのは、それが well-known で、名前から「条件が満たされなければ止まる・以降は前提が成立している」と読み手が期待できるからだと思います。今回の関数は新規追加で、その共有された期待がありません。

おそらく `ensure` という名前もその一因だと思います。これだけでは error なのか fallback なのか、あるいは set まで行うのかが分からず、実装を見るまで副作用(第三引数が fallback であって、それを set すること) が読み取れません。関数化は、ある点、可読性のためのものですが、この点ではむしろ推測コストを上げていると感じます。

もし関数化するのであれば、名前に set や fixup、fallback などを入れて、副作用が明示的であってほしいです。

個人的には、この種の処理は初期化時に一度だけ行えば十分で、副作用のある処理を使いやすくする必要はなく、下の `cond` で `setq` する形で足りると思っています。(結局、 `cond` で、この関数内の `member` 的なことをするわけなので)

tooro88 replied:
大変失礼ながら、このお返事にはすぐには納得のできない部分があって、議論にお付き合いいただけるのなら議論を続けたいという気持ちがあります。ですが、それはひとまず置いておいて、副作用が問題ということでしたら、次のコードはどうでしょうか。fallback 値への変更をなくしました。これならシンプルで、私にとっては、妥協策としてでなく、単体で満足の行くコードです。

~~~elisp
(defun tcode--load-isearch ()
  "`tcode-use-isearch' の設定に応じて必要なファイルをロードする。"
  (cond ((eq tcode-use-isearch 'overwrite)
         (require 'tc-sysdep)
         (require tcode-isearch-overwrite-module))
        ((eq tcode-use-isearch 'advice)
         (require 'tc-ishelper))
        ((eq tcode-use-isearch nil) )  ; do nothing
        (t
         (lwarn 'tcode :warning "Invalid value `%s' for `tcode-use-isearch'."
                tcode-use-isearch))))
~~~
ひとまず、以下の部分は置いておいて、このコードで ok かだけお知らせいただけないでしょうか。

以下はいただいたお返事に対する私の主張です。もし上のコードで問題無い(または微修整だけで済む)なら、以下をお読みいただく優先度は低いです。今後もこういう観点で意見がわかれるかも、ぐらいの参考情報です。コメントいただけたらありがたいですし、「へえー」と言って閉じていただいても構いません。

> ここはかなり悩ましいのですが、 `assert` が成り立つのは、それが well-known で、名前から「条件が満たされなければ止まる・以降は前提が成立している」と読み手が期待できるからだと思います。今回の関数は新規追加で、その共有された期待がありません。

新規コードに共有された知識が無いのは当然なので、この話はよく分かりません。組み込み関数以外でも、エラー終了や変数の書き換えを目的とする関数をどんどん新規に作ってよいと私は考えます。

> おそらく `ensure` という名前もその一因だと思います。これだけでは error なのか fallback なのか、あるいは set まで行うのかが分からず、実装を見るまで副作用(第三引数が fallback であって、それを set すること) が読み取れません。関数化は、ある点、可読性のためのものですが、この点ではむしろ推測コストを上げていると感じます。

コードを読んでいて、知らない関数が現れたら、ドキュメントや本体を読むべきだと私は考えます。「推測コスト」とありますが、推測で済ませてはいけないと考えます。

ドキュメントや本体を読んだ上で、名前が misleading だという指摘はあり得ると思います。`tcode--ensure-valid-value` が misleading だとは思っていませんでした。もちろん、もっといい名前を提示されたら、「そっちの方が全然いいじゃん」となった可能性は大いにあったと思います。

> もし関数化するのであれば、名前に set や fixup、fallback などを入れて、副作用が明示的であってほしいです。

もう `tcode--ensure-valid-value` は無しの案を出したのでこれ以上この関数を擁護しても意味が無いのですが、これについては、副作用の何を問題視しているのかがよくわかりません。

一般論として、変数の値がこっそり書き換わるのはよくないと思います。ですが、それは valid な値が valid な値に変わる場合ではないでしょうか。「invalid な値であることを仮定してコードを書いていたが、valid な値に変わっていて困った」とはならないはずです。(しかも、関数の先頭に書かれている処理です。) さらに、前にも書いたように、値を書き換えることが `tcode--ensure-valid-value` の目的ではありません。`tcode-use-isearch` が `nil` `overwrite` `advice` 以外の値でないことを保証すること(C の enum のようなアイデアを、lisp で表現すること)がこの関数の目的であって、それをエラー終了によって実現するのか、fallback 値への書き換えによって実現するのかは「実装の詳細」と言っていいぐらいです(ドキュメントに FALLBACK 値をセットすると書いている以上「実装の詳細」は言い過ぎですが、もっとぼかしたドキュメントになっていて、「FALLBACK 引数は現在のところ使われていない」となっていても用途が達成できるぐらいには詳細です。)

別の観点として、`tcode--ensure-valid-value` が副作用をおこすのがまずくて、`tcode--load-isearch` で副作用をおこしてもよいのはなぜか。この質問の答えが浮かばないぐらい、私は副作用のまずさがわかっていません。

> 
> 個人的には、この種の処理は初期化時に一度だけ行えば十分で、副作用のある処理を使いやすくする必要はなく、下の `cond` で `setq` する形で足りると思っています。(結局、 `cond` で、この関数内の `member` 的なことをするわけなので)

「副作用のある処理を使いやすくする必要はなく」の部分は、よくわかりません。次の疑問が浮かびます。
 - 繰り返しですが副作用のどんな点を問題視しているのかがよくわかりません。emacs lisp は基本的には副作用で処理を進めていく言語ではないでしょうか。今後のレビュー対象も、副作用だらけです。もし今後、副作用に関する問題があったときは、どういう点を問題視しているのか少し詳しめに教えてほしいです。
 - 「使いやすく」は、関数化することを指しているのでしょうか。確かに、処理を1回書くだけで何度も使えて便利、というのは関数の利点の一つです。しかし、私は、その利点のために関数化がしたかったわけではありません。私は「作者のアイデアがよく表現されているプログラムがよいプログラムである」と教わり、自分でもそれは正しいと思っています。関数は、その方針でプログラムを書くための基本的な道具です。1つの関数を複数個所から呼ぶ例で言うと、単に同じことを何度も書く手間が省けていること以上に、これら複数個所でやっている処理が「まさに同じ処理である」ということが表現されていることが重要なメリットであると考えます。今回の例では(小さいコードなのでありがたみが薄いですが、もっと複雑なコードに敷衍して考えてほしいです)、`tcode--ensure-valid-value` を関数化することによって、`tcode--load-isearch` の残りの部分とは独立した処理であることが表現されます。利点として例えば、`tcode--ensure-valid-value` に入った修正を確認するときに、`tcode--load-isearch` についての詳しい知識がいらないことがはっきりしています。一つの関数内に書き下してあったらそうはいきません。

   というわけで、私は関数化を「良いこと」ととらえているので、それを止めるように言われて驚いている次第です。(もちろん、関数化を上手くやれないないので戻した方がマシという指摘はあると思います。今回挙げていただいた理由では、上手くやれていないとは思えませんでした。)

naota replied:
> 大変失礼ながら、このお返事にはすぐには納得のできない部分があって、議論にお付き合いいただけるのなら議論を続けたいという気持ちがあります。ですが、それはひとまず置いておいて、副作用が問題ということでしたら、次のコードはどうでしょうか。fallback 値への変更をなくしました。これならシンプルで、私にとっては、妥協策としてでなく、単体で満足の行くコードです。
> 
> ```emacs-lisp
> (defun tcode--load-isearch ()
>   "`tcode-use-isearch' の設定に応じて必要なファイルをロードする。"
>   (cond ((eq tcode-use-isearch 'overwrite)
>          (require 'tc-sysdep)
>          (require tcode-isearch-overwrite-module))
>         ((eq tcode-use-isearch 'advice)
>          (require 'tc-ishelper))
>         ((eq tcode-use-isearch nil) )  ; do nothing
>         (t
>          (lwarn 'tcode :warning "Invalid value `%s' for `tcode-use-isearch'."
>                 tcode-use-isearch))))
> ```
> 
> ひとまず、以下の部分は置いておいて、このコードで ok かだけお知らせいただけないでしょうか。

これはOKだと思います. やるなら, ここで overwriteのbranchを後に持ってきて

```elisp
(cond
 ...
 (t
  (unless (eq tcode-use-isearch 'overwrite)
    (lwarn ...)
    (setq tcode-use-isearch 'overwrite))
  (require 'tc-sysdep)
  (require tcode-isearch-overwrite-module)))
```

と (advice など以外の) non-nil value を fallback してもいいと思います.

naota replied:
前提として、関数化や副作用そのものに反対しているわけではありません. 今回の懸念は、「任意のグローバル変数を書き換えうる汎用関数」を追加することによる、
変更地点の不可視化とメンテナンス性の低下です. この部分, 副作用という言い方が曖昧であったかと思います. 申し訳ないです.

一般的な話として, もちろんコードの意味のあるまとまりを関数にするのはいいことです. しかし, そこには「なにを」「どのように」関数としてまとめるのかという評価も伴います.

まず, 「なにを」の視点からいうと, わたしは, グローバルな変数を非明示的に書きかえるルートを追加する関数について, 懸念を持ちます. これは変数がどこで書きかえられているのかを, 見にくくし, メンテナンス性を下げると思うからです.

その点, 上記コメントのように明示的に `(setq tcode-use-isearch ...)` するのは問題ないと思います. なぜなら, ここでは明らかに `tcode-use-isearch` がこの場で書きかえられているのだということがわかります. 極論, ざっくりと `grep "setq tcode-use-isearch"` とすれば, 全ての変更地点を見つけることができるでしょう.

一方, 関数が受け取ったシンボルに `set` する場合, そこで何が書きかえられるのかは, 全ての呼び出しをチェックする必要があります. 今回はまだ1つの関数ですが, 複数の `set` する関数があったらどうでしょうか? それぞれの関数の呼び出しをチェックしていくことになります. そういったメンテナンス性の低下をおそれて, こういう関数の追加には懐疑的です.

もちろん, 今回の関数も上手に扱えば便利なものだとは思います. しかし, 便利であるがゆえに, 将来意図せず利用箇所が増え, 結果として変数の変更地点が見えにくくなることを危惧しています.

> 以下はいただいたお返事に対する私の主張です。もし上のコードで問題無い(または微修整だけで済む)なら、以下をお読みいただく優先度は低いです。今後もこういう観点で意見がわかれるかも、ぐらいの参考情報です。コメントいただけたらありがたいですし、「へえー」と言って閉じていただいても構いません。
> 
> > ここはかなり悩ましいのですが、 `assert` が成り立つのは、それが well-known で、名前から「条件が満たされなければ止まる・以降は前提が成立している」と読み手が期待できるからだと思います。今回の関数は新規追加で、その共有された期待がありません。
> 
> 新規コードに共有された知識が無いのは当然なので、この話はよく分かりません。組み込み関数以外でも、エラー終了や変数の書き換えを目的とする関数をどんどん新規に作ってよいと私は考えます。
> 
> > おそらく `ensure` という名前もその一因だと思います。これだけでは error なのか fallback なのか、あるいは set まで行うのかが分からず、実装を見るまで副作用(第三引数が fallback であって、それを set すること) が読み取れません。関数化は、ある点、可読性のためのものですが、この点ではむしろ推測コストを上げていると感じます。
> 
> コードを読んでいて、知らない関数が現れたら、ドキュメントや本体を読むべきだと私は考えます。「推測コスト」とありますが、推測で済ませてはいけないと考えます。

ここは「どのように」に関わる問題かと思います. もちろん, 知らない関数についてドキュメントや本体を読むべきであるのは正当です.

しかしながら, このコードはボランティアベースのOSSです. できるだけカジュアルにぱっと読んで大枠が分かることも大事にしたいです. もちろんコードを変更する人はどうせ中身まで読まなければならないとしても, コードをまずは読みたい人が, 簡単に読めることもOSSでは大切だと思っています. なので小さいコードであればあるほど, well-known であったり, 期待が明白であってほしいです. ここはある種の思想かもしれませんが, 重要な設計判断だと考えています.

> > もし関数化するのであれば、名前に set や fixup、fallback などを入れて、副作用が明示的であってほしいです。
> 
> もう `tcode--ensure-valid-value` は無しの案を出したのでこれ以上この関数を擁護しても意味が無いのですが、これについては、副作用の何を問題視しているのかがよくわかりません。
> 
> 一般論として、変数の値がこっそり書き換わるのはよくないと思います。ですが、それは valid な値が valid な値に変わる場合ではないでしょうか。「invalid な値であることを仮定してコードを書いていたが、valid な値に変わっていて困った」とはならないはずです。(しかも、関数の先頭に書かれている処理です。) さらに、前にも書いたように、値を書き換えることが `tcode--ensure-valid-value` の目的ではありません。`tcode-use-isearch` が `nil` `overwrite` `advice` 以外の値でないことを保証すること(C の enum のようなアイデアを、lisp で表現すること)がこの関数の目的であって、それをエラー終了によって実現するのか、fallback 値への書き換えによって実現するのかは「実装の詳細」と言っていいぐらいです(ドキュメントに FALLBACK 値をセットすると書いている以上「実装の詳細」は言い過ぎですが、もっとぼかしたドキュメントになっていて、「FALLBACK 引数は現在のところ使われていない」となっていても用途が達成できるぐらいには詳細です。)
> 
> 別の観点として、`tcode--ensure-valid-value` が副作用をおこすのがまずくて、`tcode--load-isearch` で副作用をおこしてもよいのはなぜか。この質問の答えが浮かばないぐらい、私は副作用のまずさがわかっていません。

この部分は, わたしの「副作用」という言い方がよくなかったと思います. 上記のとおり, 任意の変数をとり, それを変更しうるのが「便利すぎる」ことに懸念があります.

この点でいうと, もしこの関数が以下のように `tcode-use-isearch` 専用であれば問題を感じないです.

```elisp
(defun tcode--ensure-use-isearch-value ()
  "..."
  (unless (memq tcode-use-isearch '(overwrite advice nil))
    (lwarn 'tcode :warning
           "Invalid value `%s' for `tcode-use-isearch'; falling back to `overwrite'."
           tcode-use-isearch)
    (setq tcode-use-isearch 'overwrite)))
```

これは, ここで `tcode-use-isearch` を変更していることが, 関数の呼び出し元を見なくても明白だからです. また valid な value がこの関数内に閉じているのも, 良いと感じています. 複数の呼び出しがあって, valid value がそれぞれ違う(変更し忘れなどで)ことを防げます.

>     * 「使いやすく」は、関数化することを指しているのでしょうか。確かに、処理を1回書くだけで何度も使えて便利、というのは関数の利点の一つです。しかし、私は、その利点のために関数化がしたかったわけではありません。私は「作者のアイデアがよく表現されているプログラムがよいプログラムである」と教わり、自分でもそれは正しいと思っています。関数は、その方針でプログラムを書くための基本的な道具です。1つの関数を複数個所から呼ぶ例で言うと、単に同じことを何度も書く手間が省けていること以上に、これら複数個所でやっている処理が「まさに同じ処理である」ということが表現されていることが重要なメリットであると考えます。今回の例では(小さいコードなのでありがたみが薄いですが、もっと複雑なコードに敷衍して考えてほしいです)、`tcode--ensure-valid-value` を関数化することによって、`tcode--load-isearch` の残りの部分とは独立した処理であることが表現されます。利点として例えば、`tcode--ensure-valid-value` に入った修正を確認するときに、`tcode--load-isearch` についての詳しい知識がいらないことがはっきりしています。一つの関数内に書き下してあったらそうはいきません。
>       というわけで、私は関数化を「良いこと」ととらえているので、それを止めるように言われて驚いている次第です。(もちろん、関数化を上手くやれないないので戻した方がマシという指摘はあると思います。今回挙げていただいた理由では、上手くやれていないとは思えませんでした。)

コードはコードとして残るので, 意図と離れた使われ方をすることがあります. 今回, 意味のまとまりのため(だけ)に関数として作ったとしても, それは他から使いやすく使われていくことを妨げることはできません. set している場所から, なにが変更されているのか見えにくい, 変数名でgrepしてもすぐにはそれが setq とは見えにくい…というのをわたしはやはり危惧しています. これが関数として, より大きなコードであり, 複雑な処理をして, 呼び出し側で処理しにくいのであれば…まだわかるのですが, 今回のようにコードが小さく, 代替手段も明確に存在する局面では, 関数化による表現上の利点よりも、将来のメンテナンスコストの方を重く考えてしまいます.

tooro88 replied:
議論にお付き合いいただきありがとうございます。納得しました。

まず、コードは fallback 値への書き替えの無いバージョンの commit をもう作ってしまったので、このままでお願いします。(ユーザーの設定ミスはすぐに直してもらった方がよいと考えると、こちらの方がよい気もして来ました。)

`set` の使用が問題である(標語として「grep しにくくなる」)という観点は非常に明快で、私はその観点がすっぽり抜け落ちていました。言われて見ると積極的に賛同したい考え方です。(そういえば、「オブジェクト指向って grep で追跡しにくくていやだなあ」と以前考えていたこともあったのを思い出しました。)

挙げていただいた `tcode--ensure-use-isearch-value` のような関数なら問題無いというのも朗報で、それなら(私から見ると)気にすべき意見の相異は無いと言えそうです。

> また valid な value がこの関数内に閉じているのも, 良いと感じています. 複数の呼び出しがあって, valid value がそれぞれ違う(変更し忘れなどで)ことを防げます.

これはおっしゃる通りで、私が主張していた路線に沿うなら、むしろ私が率先してこう書いているべき(`tcode--ensure-valid-value` よりもまず先に `tcode--ensure-use-isearch-value` があるべき)でした。なるほど。

ほかの点も同意です。返信にずいぶんお手間をとらせてしまったと思いますが、おかげで納得のできる結論となりました。ありがとうございました。

naota commented on tc-ishelper.el:
```diff
@@ -21,4 +21,76 @@
+    (funcall orig-fun char count)))
+
+;; tc-is22.el の isearch-printing-char の追加部分を抜き出したもの。
+(defun tcode--input-method-for-isearch (char count)
```
ここ実は `count` がちゃんと処理されていない気はするのですが、まあ実害?はないんですかね。まあいつか直せばいいのかな…。

tooro88 replied:
これはバグです。tc-is22.el に元からある問題であり、advice 実装でも、`tcode--input-method-for-isearch` が元の tc-is22.el のコードと字面レベルで変更が無いことを優先して、放置してしまいました。(言及ぐらいすべきでしたが、忘れていました。)

次のような状況です。
 - 素の Emacs では、
    * `C-s C-u a` で `aaaa` がサーチ文字列となる。 
 - tc-is22.el の実装
    * `C-s C-u a` で `a` がサーチ文字列となる。(バグ)
    * `C-s C-u あ` で `あ` がサーチ文字列となる。(バグ)
 - advice 実装
    * `C-s C-u a` で `aaaa` がサーチ文字列となる。(正常)
    * `C-s C-u あ` で `あ` がサーチ文字列となる。(バグ)
 - im 実装
    * `C-s C-u a` で `aaaa` がサーチ文字列となる。(正常)
    * `C-s C-u あ` で `ああああ` がサーチ文字列となる。(正常)

直すのは(予想では)簡単ですが、この PR に含めましょうか?(先にtc-is22.elを修正して、advice版はそれのコピーとする) それとも、全部済んでからにしましょうか?(両者別々に修正する) レビューしやすいのはどちらですか?

naota replied:
このPR自体長くなっているので、もともとバグなら別件でよいかと思います。

tooro88 replied:
了解です。Issue #42 に問題を書き写しておきました。

tooro88 commented:
emacs-24、25 をサポート外にしたのはよいのですが、今の修正方法だとそれらを起動して tc-setup を読み込んだだけでエラーになってしまいます(存在しないマクロ isearch-define-mode-toggle を使用している)。それはちょっと不便に感じたので、エラーにはならないように直します。

tooro88 committed:
Remove unused code from tc-is22.el

`isearch-toggle-tcode` is only referenced in obsolete XEmacs-specific
code.  Remove both.

tooro88 committed:
Rename t to 'overwrite for tcode-use-isearch

Upcoming changes will provide multiple implementations of the Isearch
extensions.  The variable `tcode-use-isearch` will be used to select
among them.

Rename the default value `t` to `'overwrite`, which more accurately
describes the default implementation.

tooro88 committed:
Clarify and fix the behavior of tcode-isearch-start-state

 - Clarify the variable's docstring to better describe its behavior.
 - Remove the redundant value t.
 - For values 0 and 1, fix incomplete handling of input method state
   changes: update the mode line and maintain consistent internal
   state by using `isearch-toggle-input-method`.

tooro88 committed:
Rename "wrapped search" to "line-fold search"

The term "wrapped search" in tc-is22.el is confusing.  It refers to a
search that ignores line wrapping, while Isearch already uses "wrap"
to describe restarting a search from the opposite end of the buffer.

Rename "wrapped search" in tc-is22.el to "line-fold search" to avoid
this ambiguity.

tooro88 committed:
Rename tcode-isearch-type

Rename the variable `tcode-isearch-type` to
`tcode-isearch-overwrite-module`, which more accurately describes its
role.

tooro88 committed:
Split tc-is22.el

Split tc-is22.el into two files for future reuse:
 - Keep in tc-is22.el only the functions that override Emacs internal
   functions.  These will be reimplemented later.
 - Move the remaining functions to tc-iscommon.el, which will be
   reused in the forthcoming reimplementation.

tooro88 commented:
レビュー結果を反映しました。以下が内容に変更のある commit です。

 - (commit a72cd69) Rename "wrapped search" to "line-fold search"
    + 新規。tc-is22.el での `wrapped` → `line-fold` の置換
    + `tcode-isearch-enable-wrapped-search` の obsolete 化
 - (commit a5a8246) Add 'advice option to tcode-use-isearch
    + `tcode-use-isearch` の doc-string を更新
 - (commit 9ca918d) Reimplement Isearch extensions using advice
    + `wrapped` → `line-fold` の置換
 - (commit f716bed) Reimplement line-fold search using isearch-regexp-function
    + `wrapped` → `line-fold` の置換
    + emacs-24,25 でエラーにならないよう修正
 - (commit bb33e98) Add ERT tests for Isearch extensions
    + isearch 中の prefix argument のテスト追加
    + `tcode-electric-space` のテスト追加
    + `:expected-result` に github issue へのリンクのコメント追加
    + コンパイラ warning 対策、不要行の整理、コメントの表記統一などマ
      イナー修正

naota commented on test/tctest-env.el:
```diff
@@ -0,0 +1,139 @@
+;;   isoverwrite : tcode-use-isearch を 'overwrite に設定する。
+;;   isadvice    : tcode-use-isearch を 'advice    に設定する。
+;;   isim        : tcode-use-isearch を 'im        に設定する。
+;;   notc        : tc-setup をロードしない。
```
`ist` が抜けてそうです。これらの値は排他ということでよいでしょうか?

tooro88 replied:
`ist` は undocumented な隠しフィーチャーのつもりです。commit 3be2b29 (Rename t to 'overwrite for tcode-use-isearch) よりも手前の commit をテストするときに私が使います。

現状の実装は排他ですが、ここでドキュメントに書いて決めてしまう必要はないと考えます。`notc` だけ排他にせず、起動時は `notc` だけども後から手動で tc-setup.el をロードしたときに、`advice` として動く、ということを可能にすることも十分考えられます。

naota replied:
テストは開発者に向けたものなので, あまり隠すこともないような…まあわかりました.

排他かどうか気になったのは, `tctest-env-load-tc` の `when` 部分があったからです. そちらにdocuemntが入ったので, まあよいかなと思います.

naota commented on test/tctest-env.el:
```diff
@@ -0,0 +1,139 @@
+      (when (member w keywords)
+	(setq count (1+ count))))
+    (when (/= count 1)
+      (error "One of %s must be specified as keyword"
```
"One of"だといずれか1つ(複数でもよい)になるので、"Exactly one of"ですかね

naota replied:
dolistの部分のearly returnも少し考えましたが…まあそこまでするでもないですね…

tooro88 replied:
> "One of"だといずれか1つ(複数でもよい)になるので、"Exactly one of"ですかね

たしかに、"Exactly one of" ですね。直します。

> dolistの部分のearly returnも少し考えましたが…まあそこまでするでもないですね…

2個 hit したら、その時点で残りを調べる必要がなくなる、ということでしょうか。まあ必要ないと思います。

naota commented on test/tctest-env.el:
```diff
@@ -0,0 +1,139 @@
+      (insert msg "\n"))))
+
+(defun tctest-env-load-tc (keywords)
+  "変数設定をした上で、tc-setup をロードする。"
```
以下のwhenは、keyword checkされているので、どれか1つしかhitしないということですね。その点、コメントを入れておきたいです。

tooro88 replied:
はい、コメントたしておきます。

naota commented on test/tctest-env.el:
```diff
@@ -0,0 +1,139 @@
+
+(defun tctest-env-main ()
+  (let* ((keywords-packed (or (getenv "TCTEST_ENV")
+			   (error "TCTEST_ENV not specified")))
```
indentが少しずれていそうです

tooro88 replied:
ほんとだ…直します。

naota commented on test/tctest-env.el:
```diff
@@ -0,0 +1,139 @@
+	     (not (member "nobatch" keywords)))
+    ;; FIXME: quiet オプションは、実験的に付けてはみたものの、isearch
+    ;; の echo area 表示が大量に出力されるので、あまり意味はなかった。
+    ;; echo area 表示の stderr への出力を止める方法はある?
```
たとえばこうですかね? lexical-bindingでも動くのかな…?

```elisp
(flet ((message (&rest args)))
  (message "foo"))
```

tooro88 replied:
試したところ、flet (cl-flet ですよね?)だと isearch の表示は消せませんでした。(isearch.el が lexical-binding だからかも。よく調べてません。) ですが、`(let ((inhibit-message t)) ...` という機能があるようで、これで消せそうです。

ただ、この位置でやると ert の出す summary も無くなってしまうんですよね。個々のテストの中でだけ消えるようにする機能追加が必要です。追い追い検討します。

naota replied:
はい, 特に急ぐ理由もないので後でよいと思います.

naota reviewed (commented):
テスト部分まで全部見ようかと思ったのですが…今日はここまでですみません。明日また見ます。

tooro88 commented:
> テスト部分まで全部見ようかと思ったのですが…今日はここまでですみません。明日また見ます。

ごゆっくりで全然構いません。私の方には全く急ぐ理由がなくて、見てもらえているだけでありがたいです。

よく考えると、テストの部分に直しがあってもそれ以前の commit に影響があるとは考えにくいので、別 PR にした方が良い気がしてきましたが、どうでしょうか。自分で一緒にしておいてなんですが。

naota commented:
> よく考えると、テストの部分に直しがあってもそれ以前の commit に影響があるとは考えにくいので、別 PR にした方が良い気がしてきましたが、どうでしょうか。自分で一緒にしておいてなんですが。

基本的には変更とテストは同時に入るべきですね。ただ今回、テストの基本的な機能自体の実装なので別PRでもよいといえばよいですね。

まあ分けた方が若干このPRのマージは早くなるかとは思いますが…もう見始めてはいるのであまり変わりはしないかも…

tooro88 commented:
> 基本的には変更とテストは同時に入るべきですね。

なるほど。ではこのままで行きましょう。

tooro88 committed:
Add 'advice option to tcode-use-isearch

Add an `'advice` option to `tcode-use-isearch`.  When this option is
selected, the loader skips tc-is22.el and instead loads
tc-ishelper.el, which will reimplement the features of tc-is22.el
using advice in forthcoming commits.

tooro88 committed:
Reimplement Isearch extensions using advice

The Isearch extensions provided by tc-is22.el override several
internal Emacs functions.  This approach is error-prone and can miss
new features introduced in recent Emacs versions.

tc-is22.el provides two features:
 1. line-fold search, which will be reimplemented in the next commit, and
 2. user input hooking, which is reimplemented here.

Replace the modification of `isearch-printing-char` with advice that
achieves the same behavior -- invoking the T-Code input method when
necessary -- so that the existing bushu/mazegaki conversion code in
tc-iscommon.el continues to work unchanged.

tooro88 committed:
Add a section about Isearch extensions to README.md

tooro88 committed:
Reimplement line-fold search using isearch-regexp-function

tc-is22.el implements line-fold search by modifying internal Isearch
functions to convert a search string into a regexp.  Since Emacs 25,
`isearch-regexp-function` has provided a cleaner way to perform such
conversions.

Use `isearch-regexp-function` to reimplement line-fold search, and
enable this implementation when `tcode-use-isearch` is `'advice`.
Together with the change in the previous commit, this completes the
reimplementation of the features previously provided by tc-is22.el.

tooro88 commented:
レビュー結果を反映しました。変更のあったcommit:

 - (commit 48e221c) Add 'advice option to tcode-use-isearch
    * `tcode--load-isearch` の、上のコメントに書いた版を入れました(コメントでまだokもらってませんが、仮の案ということで)。
 - (commit 8e41e06) Add ERT tests for Isearch extensions
    * 「One of」 → 「Exactly one of」
    * `tctest-env-load-tc` へのコメント追加
    * `tctest-env-main` のインデント修正
 - (commit c85b5fa) Implement quiet mode for tctest
    * テストのquiet機能です。必須の修正ではないですが、思いのほかいい感じで、あった方が便利なので、追加しました。

tooro88 commented:
関数名、変数名の `--` について。

当初、tctest-play.el で定義する名前について、「内部関数は `tctest--`、外部に見せる関数は `tctest-` で始める」という方針で書き始めたのですが、今見直すと、どうも徹底できておらず、テストの何が外部で何が内部かの規準もいまいちよくわかりません。

いっそ内部外部の区別をやめて、全部 `tctest-` で統一してしまう方が、今後のコード修正の際も余計なことで悩まずすっきりすると思うのですが、どうでしょうか。

ok なら、レビューのキリのいいタイミング(こちらが修正するターンになったとき)にまとめて変更します。または、「区別する路線を徹底するべき」「ok だがこの PR の後で」などご意見あればお知らせください。

naota commented:
> いっそ内部外部の区別をやめて、全部 `tctest-` で統一してしまう方が、今後のコード修正の際も余計なことで悩まずすっきりすると思うのですが、どうでしょうか。

年末年始なかなかreviewできずすみません…

どうせemacsでは内外を厳密に区別することもできないですし, package全体としても特別の区別も現状できていませんし, この方針で問題ないかと思います.

tooro88 commented:
助かります。次回何かを修正するタイミングで一緒に直します。

naota commented on test/tctest-play.el:
```diff
@@ -0,0 +1,440 @@
+	(tctest-define-key params (symbol-value key-map) (kbd key) cmd)))))
+
+(defun tctest-where (cmd)
+  "コマンド CMD に割り当てられた T-Code キー列(文字のリスト)。"
```
ここですが, コメントになにを与えてなにが返るのかのexampleがあるとよいかなと思いました.
また, 現状だと CMD (symbol)をとるようですが, これはtcode-tableを引くより, tc-tbl.elを読んだ上で `tcode-special-commands-alist` をひいてもいいのかな?と思いました,

tooro88 replied:
はい、例を書いておきます。

`tctest-where` の実装ですが、現状の `tcode-table` を引く作戦のままでいきたいです。というのも、今、手元でテストを修正していて、TUT-Code などでもテストが通るようにしているところです。TT-Code や Try-Code では、表が2段構えになっていて、複数の *tbl.el のそれぞれで `tcode-special-commands-alist` を定義しています。たとえば、TT-Code では、普段は ttc-tbl.el の表が使われるのですが、`fj` を押すと、ttc-ltbl.el の表が有効になり、そのファイルの `tcode-special-commands-alist` はその表専用の定義になっています。これを読む実装はややこしくなりそうです。`tcode-table` を引くなら、その辺を気にしないで済みます。(今の実装を書いたときにそこまで見通していたわけではないのですが。)

naota replied:
なるほど, わかりました.

naota commented on test/tctest-play.el:
```diff
@@ -0,0 +1,440 @@
+  "文字列 KEYS の中の T-Code 文字と特殊キーワードをキー列に置換する。"
+  (tctest--untc (tctest--expand-key-abbrevs keys)))
+
+(defun tctest-show-modeline ()
```
tctest-show-modelineは, "[]部分"と言っていますが, ちょっとあいまいなので current-input-method-title か, IMEのタイトルなどと言ってよいと思います.

tooro88 replied:
確かに変なドキュメントですね。直します。

naota commented on test/tctest.el:
```diff
@@ -0,0 +1,998 @@
+  ;; その他のキーワード:
+  ;;   [POSTBUSHU] [PREBUSHU] [POSTMAZE] [PREMAZE] : 部首変換/交ぜ書き変換
+  ;;   [ALNUM_ZEN] [ALNUM_HAN] : 全角英数/半角英数切り換え
+  ;;   [KUTEN_J] [KUTEN_A]     : 日本語句読点/ASCII句読点切り換え
```
これらの特殊キーワードと"#"でのカタカナ変換ふくめて, 冒頭に(も)ドキュメントを持ってくると, ぱっと見でもわかってよさそうに思います.

tooro88 replied:
その方が良いですね。直します。

naota reviewed (commented):
いくつかコメントを書きましたが, 基本的にはoptionalです. 全体的にはもう完成でマージできるものと思います.
上の `tctest--` の変更をいれていただいて, あとはマージでよさそうと思っています.

tooro88 commented:
> いくつかコメントを書きましたが, 基本的にはoptionalです.

せっかくなので、コメントいただいた分も反映したいと思います。

ついでに次の2点も直させてください。どちらも一行程度の修正です。
 - `tctest-play` の docstring に `:quiet` の説明を追加し忘れていた点。
 - テストアイテム `tctest-katakana-mode` で、カタカナモードのテストなのにカタカナを直接入力してしまっている点。

tooro88 committed:
Implement quiet mode for tctest

Add `:quiet` option to tctest-play and enable it when `quiet` keyword
is specified for tctest-env.

tooro88 committed:
Add ERT tests for Isearch extensions

New files:
 - tctest.el      : Test items.
 - tctest-play.el : Test driver.  Simulate user inputs using keyboard macro.
 - tctest-env.el, run.bash : Convenient scripts for selecting isearch options.

tooro88 commented:
レビュー結果を反映しました。追加で、`tctest-cmp` 用のキーワード引数の説明が抜けていたので docstring に追加しました。

変更したcommit:
 - (commit 8bec74b) Add ERT tests for Isearch extensions
    * `tctest-where` にコメントで例を追加。docstring の返り値の説明が間違っていたので修正。
    * `tctest-show-modeline` の docstring を修正。
    * `tctest--` → `tctest-`
    * `tctest-cmp` の docstring にキーワード引数の説明を追加。
    * tctest.el の冒頭に特殊キーワードとカタカナモードの説明を追加。
    * テストアイテム `tctest-katakana-mode` にて、「ア」の直接入力を「あ」に
 - (commit 87bd037) Implement quiet mode for tctest
    * `tctest-play` の docstring に `:quiet` の説明を追加。

naota commented:
マージしました! 長きにわたりありがとうございました.
